cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -535,6 +540,138 @@ public void programSpecification(HttpRequest request, HttpResponder responder,
+    if (scheduleSpecFromRequest.getSchedule().getName() != null &&
+      !scheduleName.equals(scheduleSpecFromRequest.getSchedule().getName())) {
+      throw new BadRequestException(
+        String.format(
+          "schedule name in the body of the request (%s) does not match the schedule name in the path parameter (%s)",
+          scheduleSpecFromRequest.getSchedule().getName(), scheduleName));
+    }
+
+    lifecycleService.addSchedule(applicationId, scheduleSpecFromRequest);
+    responder.sendStatus(HttpResponseStatus.OK);
+  }
+
+  @POST
+  @Path("apps/{app-name}/schedules/{schedule-name}/update")
+  @AuditPolicy(AuditDetail.REQUEST_BODY)
+  public void updateSchedule(HttpRequest request, HttpResponder responder,
+                             @PathParam("namespace-id") String namespaceId,
+                             @PathParam("app-name") String appName,
+                             @PathParam("schedule-name") String scheduleName)
+    throws SchedulerException, BadRequestException, NotFoundException, IOException, AlreadyExistsException {
+    doUpdateSchedule(request, responder, namespaceId, appName, ApplicationId.DEFAULT_VERSION, scheduleName);
+  }
+
+  @POST
+  @Path("apps/{app-name}/versions/{app-version}/schedules/{schedule-name}/update")
+  @AuditPolicy(AuditDetail.REQUEST_BODY)
+  public void updateSchedule(HttpRequest request, HttpResponder responder,
+                             @PathParam("namespace-id") String namespaceId,
+                             @PathParam("app-name") String appName,
+                             @PathParam("app-version") String appVersion,
+                             @PathParam("schedule-name") String scheduleName)
+    throws SchedulerException, BadRequestException, NotFoundException, IOException, AlreadyExistsException {
+    doUpdateSchedule(request, responder, namespaceId, appName, appVersion, scheduleName);
+  }
+
+  // TODO: refactor doAddSchedule and doUpdateSchedule to reduce code duplication
+  private void doUpdateSchedule(HttpRequest request, HttpResponder responder, String namespaceId, String appId,
+                                String appVersion, String scheduleName)
+    throws IOException, BadRequestException, NotFoundException, SchedulerException, AlreadyExistsException {
+    ApplicationId applicationId = new ApplicationId(namespaceId, appId, appVersion);
+    ScheduleSpecification scheduleSpecFromRequest;
+    try (Reader reader = new InputStreamReader(new ChannelBufferInputStream(request.getContent()), Charsets.UTF_8)) {
+      // The schedule spec in the request body can contain only the changed information or more, but should always
+      // contain the schedule type for deserialization
+      scheduleSpecFromRequest = GSON.fromJson(reader, ScheduleSpecification.class);
+    } catch (IOException e) {
+      LOG.debug("Error reading schedule specification from request body", e);
+      throw new IOException("Error reading request body");
```

Comment:
The exception message is sent as the error string along with the HTTP status code to the user. Hence we use a user friendly message here.